src/main/java/com/amazon/ion/impl/IonRawTextWriter_1_1.kt

```diff
@@ -0,0 +1,551 @@
+            Struct -> '}'
+            SExp, EExpression -> ')'
+            List -> ']'
+            Top -> throw IonException("Nothing to step out of.")
+        }
+
+        currentContainer = ancestorContainersStack.removeAt(ancestorContainersStack.size - 1)
+
+        closeValue {
+            if (options.isPrettyPrintOn && currentContainerHasValues) {
+                output.appendAscii(options.lineSeparator())
+                output.appendAscii(" ".repeat(ancestorContainersStack.size * 2))
+            }
+            output.appendAscii(endChar)
+        }
+    }
+
+    override fun writeTaggedPlaceholder() {
+        writeScalar { output.appendAscii("(:?)") }
+    }
+
+    override fun writeTaggedPlaceholderWithDefault(default: Consumer<IonRawWriter_1_1>) {
+        writeScalar {
+            stepInSExp(false)
+            output.appendAscii(":?")
+            isPendingSeparator = true
+            default.accept(this)
+            stepOut()
+        }
+    }
+
+    override fun writeTaglessPlaceholder(taglessEncodingOpcode: Int) {
+        writeScalar {
+            output.appendAscii("(:? ")
+            val tag = TaglessScalarType.getTaglessScalarTypeForOpcode(taglessEncodingOpcode)!!
+            writePrimitiveEncodingTag(tag)
+            output.appendAscii(")")
+        }
+    }
+
+    override fun stepInDirective(directiveOpcode: Int) {
+        val directiveName = when (directiveOpcode) {
+            OpCode.DIRECTIVE_SET_SYMBOLS -> "set_symbols"
+            OpCode.DIRECTIVE_ADD_SYMBOLS -> "add_symbols"
+            OpCode.DIRECTIVE_SET_MACROS -> "set_macros"
+            OpCode.DIRECTIVE_ADD_MACROS -> "add_macros"
+            OpCode.DIRECTIVE_USE -> "use"
+            OpCode.DIRECTIVE_MODULE -> "module"
+            OpCode.DIRECTIVE_IMPORT -> "import"
+            OpCode.DIRECTIVE_ENCODING -> "encoding"
+            else -> throw IonException("Not a directive $directiveOpcode")
+        }
+        stepInSExp(false)
+        output.appendAscii(":\$ion ")
+        output.appendAscii(directiveName)
+        isPendingSeparator = true
+    }
+
+    private fun writeMacroEncodingTag(macroName: String) {
+        output.appendAscii("{:")
+        output.appendAscii(macroName)
+        output.appendAscii("}")
+    }
+
+    private fun writePrimitiveEncodingTag(taglessScalarType: TaglessScalarType) {
+        output.appendAscii("{#")
+        output.appendAscii(taglessScalarType.textEncodingName)
+        output.appendAscii("}")
+    }
+
+    override fun stepInTaglessElementList(taglessEncodingOpcode: Int) {
+        openValue {
+            output.appendAscii("[")
+            val tag = TaglessScalarType.getTaglessScalarTypeForOpcode(taglessEncodingOpcode)!!
+            writePrimitiveEncodingTag(tag)
+            output.appendAscii(" ")
+        }
+        ancestorContainersStack.add(currentContainer)
+        currentContainer = List
+        currentContainerHasValues = false // So that it does what?
```

Comment:
Is this comment a note to add a comment that explains why this line of code is necessary?

Reply:
Yes.